system/jlib/jmutex.hpp

```diff
@@ -661,35 +710,74 @@ class jlib_decl ReadWriteLock
     void lockWrite()        { lockWrite(false, 0); }
     bool lockRead(unsigned timeout) { return lockRead(true, timeout); }
     bool lockWrite(unsigned timeout) { return lockWrite(true, timeout); }
-    void unlock()           { 
-                                cs.enter(); 
-                                if (readLocks) readLocks--;
-                                else
-                                {
-                                    writeLocks--;
+    void changeToWrite()
+    {
+        changeToWrite(false, 0);
+    }
+    bool changeToWrite(unsigned timeout)
+    {
+        return changeToWrite(true, timeout);
+    }
+    bool changeToRead()
+    {
+        cs.enter();
+        if (!writeLocks)
```

Comment:
minor - this could be checked outside the critical section.  Should it be an internal error rather than returning false?

Reply:
I think a boolean result indicating not locked failure, similar to the failure on timed lock read/write calls.

>this could be checked outside the critical section

I'm not sure this edge case makes it worth moving the test, and clearer semantics if inside crit ?

Reply:
I don't really mind if it is inside or outside.
If the function should never be called in that situation then it should be an internal error.  If it is expected then there should be a comment e.g.
//If the lock is already a read lock then this function will return false.

Similarly for changeToWrite - if the lock is already a write lock then it will return false.
Is that the correct response though?  If there is a timeout you can't distinguish between timedout and couldn't get the lock, and already had the write lock.

I don't have a strong view on which semantics you choose, but it should have a comment documenting what they are.

Reply:
>If there is a timeout you can't distinguish between timedout and couldn't get the lock, and already had the write lock.

ok yes I agree, on reflection I think these unexpected use cases should throw an exception instead. Will change.